File: Question1/Swift/main.swift
func maxProfit(_ prices: [Int]) -> Int {
    let n: Int = prices.count
    if n == 1 {
        return 0
    }
    // Below is an array contains the greatest selling price right of the current index
    var sellPrices = Array<Int>(repeating: 0, count: n) 
    var highest: Int = prices[n - 1]
    sellPrices[n - 1] = highest
    
    for i in stride(from: n - 2, to: -1, by: -1) {
        if prices[i] > highest {
            highest = prices[i]
        }
        sellPrices[i] = highest
    }
    
    var maxProfit: Int = 0
    for i in 0 ..< n {
        let profit: Int = sellPrices[i] - prices[i]
        if profit > maxProfit {
            maxProfit = profit
        }
    }
    return maxProfit
}

func main() {
    let prices: [Int] = [7, 1, 5, 3, 6, 12, 2, 11, 9, 4, 10, 2, 8]
    let maxProfit: Int = maxProfit(prices)
    print(maxProfit)
}


main()

File: Question1/Rust/main.rs
fn max_profit(prices: Vec<i32>) -> i32 {
    let n = prices.len();
    let mut sell_prices = vec![0; prices.len()];
    let mut highest_price = prices[n - 1]; // Start at the very last
    sell_prices[n - 1] = highest_price;
    // Iterate backwards starting at n - 2 by using skip
    for (i, price) in prices.iter().enumerate().rev().skip(1) {
        if *price > highest_price {
            highest_price = *price;
        }
        sell_prices[i] = highest_price;
    }
    let mut max_profit = 0;
    for i in 0..n {
        let profit = sell_prices[i] - prices[i];
        if profit > max_profit {
            max_profit = profit;
        }
    }
    return max_profit;
}


fn main() {
    let prices = vec![7, 1, 5, 3, 6, 4];
    println!("{}", max_profit(prices));
}